documentation/adr/0004-trip-modifications-mqtt-topic-structure.md

```diff
@@ -0,0 +1,179 @@
+# 4. Trip Modifications MQTT Topic Structure
+
+Date: 2024-07-04
+
+## Status
+
+Draft
+
+## Context
+
+[With the creation of #ADR0003](./0003-use-mqtt-to-publish-trip-modifications.md),
+we are trying to decide on the structure for _how_ we send and persist these 
+[Trip Modifications](https://gtfs.org/realtime/reference/#message-tripmodifications) between Skate and Transit Data.
+
+We know that we'll need these messages to be available if the Transit Data
+service is restarted.
+
+## Decision
+
+We'll use a
+[topic hierarchy](https://www.hivemq.com/blog/mqtt-essentials-part-5-mqtt-topics-best-practices/)
+which allows us to manage individual GTFS-RT messages and update them, and does
+not require that the Transit Data service has some way to persist state between
+restarts.
+
+### Topic Hierarchy
+We'll nest all of our Trip Modifications related data under the
+[MQTT "topic level"](https://www.hivemq.com/blog/mqtt-essentials-part-5-mqtt-topics-best-practices/)
+`trip_modifications`, which will be under our "topic prefix"
+configured in the DevOps repo for each environment.
+
+For Example:
+
+Our `<prefix>`s would be derived from the environment and are described via IaC
+in the devops repo
+```
+skate/prod
+skate/dev
+skate/dev-blue
+skate/dev-green
+```
+And the final combined topic hierarchy would be formed by prepending _one_ of 
+the `<prefix>`s
+```
+<prefix>/trip_modifications/
+```
+For instance,
+```
+skate/dev-blue/trip_modifications/
+```
+
+
+For each new and unique trip modification, we'll generate a unique ID and use
+that as our Trip Modification identifier, named `id`.
+
+> [!NOTE]  
+> The GTFS-RT Trip Modifications spec does not currently have a ID for each modification,
```

Comment:
suggestion: the `FeedEntity` (of which `TripModification` is a child) does have an ID: https://github.com/google/transit/blob/master/gtfs-realtime/proto/gtfs-realtime.proto#L86-L95

Reply:
It's interesting that the field is not present on the actual documentation definition https://gtfs.org/realtime/reference/#message-feedentity

Reply:
It's there for me?
<img width="1230" alt="Screenshot 2024-07-16 at 1 09 09 PM" src="https://github.com/user-attachments/assets/72988654-66e9-4b33-b54c-110982c9e057">

Reply:
Oh I mean there's no `trip_modification` field.
![image](https://github.com/user-attachments/assets/6efdc52a-0735-44e2-8bf8-9f00f5e1f5ed)

Reply:
Because it is there on a different page https://gtfs.org/realtime/feed-entities/#trip-modifications

(A side note: This kinda points to why I'd like to attempt to propose to define the GTFS spec in a machine readable format so docs can be autogenerated as much as possible)

Reply:
I've updated this to use a `FeedEntity`. This now means that Skate's implementation of this and this ADR are not in sync anymore.

Reply:
Do we (should we?) have a ticket to bring the implementation in line with the ADR?

Or do we want to wait until we confirm that the ADR is fully what we want it to be?

Reply:
https://app.asana.com/0/1148853526253420/1207900514697511/f